### ISSUE TITLE
Making methods more future-proof

### ISSUE THREAD

**Issue by jakearchibald** [heycam/webidl#954]
I've recently seen a couple of code examples that could create compatibility issues later:

```js
const nextFrame = new Promise(requestAnimationFrame);
```

The above works because `requestAnimationFrame` only takes one arg. If we change it to take two, the above may behave differently and unexpectedly.

```js
const controller = new AbortController();
el.addEventListener(name, callback, controller);
```

The above works because [`AbortController`](https://dom.spec.whatwg.org/#abortcontroller) and [`AddEventListenerOptions`](https://dom.spec.whatwg.org/#dictdef-addeventlisteneroptions) share a `signal` property with the same intent. If we change either of these in a way that creates a cross-over that doesn't have the same intent, the above may behave differently and unexpectedly.

Our current strategy is to make changes to the platform, and be forced into different pattern if it creates a compatibility issue due to patterns like above. Should we do something more preemptive?

We could throw in cases where functions are called with more arguments than expected, and throw if option objects have unexpected properties (aside from the properties that already come with `Object`).

This would likely be breaking if applied to existing APIs, so it should only be applied to new APIs. Existing APIs could show a console warning instead.

**Comment by annevk** [comment]
Unless JavaScript (i.e., TC39) acts here I don't think we should.

cc @littledan

**Comment by jakearchibald** [comment]
I guess the alternative is to continue outreach like https://jakearchibald.com/2021/function-callback-risks/ and hope developers follow it.

**Comment by littledan** [comment]
It's not clear to me what JavaScript should do, beyond making sure that the safer idioms are ergonomic (which I believe they are, with arrow functions and object literals). We could add certain stricter checks to make the unsafe patterns more unlikely, but to preserve compatibility, we could really only do this for new APIs (both creating a discontinuity and leaving the greater part of the problem unsolved). I wonder if type systems can help encourage people to use future-proof patterns @rbuckton.

**Comment by jakearchibald** [comment]
> we could really only do this for new APIs (both creating a discontinuity and leaving the greater part of the problem unsolved)

Agreed, although we can use console warnings for existing APIs. We may also choose to declare "we will never add an extra param to this", meaning things like `array.filter(Boolean)` would never show a warning.

**Comment by rbuckton** [comment]
The first example is one of the motivations behind https://github.com/tc39/proposal-partial-application:

```js
const nextframe = new Promise(requestAnimationFrame(?));
```

Type systems can help in some ways (informing the user if the type of the argument is incompatible), whereas `Symbol`-based protocols can help in others. I'm not sure there's a single solution given how flexible JavaScript is.

**Comment by jakearchibald** [comment]
@littledan I'm happy to present this idea at TC39 as a discussion point if you think it'd be useful. But totally get it if you think it's a no-go.

**Comment by jakearchibald** [comment]
@littledan is this true https://twitter.com/hrvbrs/status/1357425226309599232?

**Comment by annevk** [comment]
@littledan ping.
